storage: RaftTransport only needs one stream per node

bdarnell opened the pull request:
Replace the per-replica streams with a stream per replica. Rework error handling so that we can return errors without causing the whole stream to be torn down. 

This fixes #8304 and reverses the increase in goroutines noted in #8287. Future work will use the streaming response to provide snapshot acknowledgements as a stopgap while streaming snapshots are in development.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8495)

<!-- Reviewable:end -->

bdarnell committed:
storage: Stream error responses in RaftTransport

Errors no longer cause the entire stream to be torn down.
Transport-level errors are contained within RaftTransport; only errors
that are specifically addressed to Replicas are forwarded to the Store.

bdarnell committed:
storage: Return structured errors via RaftTransport

bdarnell committed:
storage: RaftTransport only needs one queue per node

Fixes #8304

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 12 of 12 files at r1, 8 of 8 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.proto, line 198 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOx_GWSUV260oc1i0Rt:-KOx_GWSUV260oc1i0Ru:bf1y3cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/18cc7fb935fee525e04048814cedfa2aa8ff2dcc/roachpb/errors.proto#L198)):*
> ```protobuf
> }
> 
> // A ReplicaTooOldError indicates a replica that has been removed from
> ```

can you make this comment more specific by discussing its sender and recipient?

---

*[storage/raft.proto, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxXpgohG1e8X_kU4sT:-KOxXpgohG1e8X_kU4sU:bi7dkmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft.proto#L39)):*
> ```protobuf
> }
> 
> message RaftMessageResponseUnion {
> ```

what are you anticipating here?

---

*[storage/raft_transport.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxY81qJuN5YdMcB1No:-KOxY81qJuN5YdMcB1Np:bf2lin3) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L53)):*
> ```Go
> // RaftMessageHandler is the interface that must be implemented by
> // arguments to RaftTransport.Listen. If either method returns an
> // error the stream will be shut down.
> ```

this appears to be untrue of `HandleRaftRequest`.

---

*[storage/raft_transport.go, line 163 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxY0retIaVOZv_6cEu:-KOxY0retIaVOZv_6cEv:bfkbyv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L163)):*
> ```Go
> 							},
> 						}
> 						if sendErr := stream.Send(resp); sendErr != nil {
> ```

i think you can name this `err` and have it do the right thing.

irrelevant after r2.

---

*[storage/raft_transport.go, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYFm7Z9kPHufQJq46:-KOxYFm7Z9kPHufQJq47:b4p1dh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L178)):*
> ```Go
> 		return err
> 	case <-t.rpcContext.Stopper.ShouldQuiesce():
> 		return nil
> ```

I think this wants to send the "node is stopping" error.

---

*[storage/raft_transport.go, line 271 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYR-1ZqzuqenItYHV:-KOxYR-1ZqzuqenItYHW:b-eqcxya) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L271)):*
> ```Go
> 				for {
> 					var resp RaftMessageResponse
> 					if err := stream.RecvMsg(&resp); err != nil {
> ```

Why not `stream.Recv()`?

---

*[storage/raft_transport.go, line 369 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYpH30CJ_h5aB2Q3Y:-KOxYpH30CJ_h5aB2Q3Z:b7o3zbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L369)):*
> ```Go
> 		if err := t.rpcContext.Stopper.RunTask(func() {
> 			t.rpcContext.Stopper.RunWorker(func() {
> 				if err := t.processQueue(ch, conn); err != nil && !grpcutil.IsClosedConnection(err) {
> ```

you don't want to log this on nil errors?

---

*[storage/raft_transport.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYwNph6mK10dkdMYs:-KOxYwNph6mK10dkdMYt:b-qglej3) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L379)):*
> ```Go
> 			})
> 		}); err != nil {
> 			return false
> ```

perhaps `sendAsync` wants to return an error.

---

*[storage/raft_transport.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOx_e_YECMl5EAZPFDa:-KOx_e_YECMl5EAZPFDb:ba90gpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/18cc7fb935fee525e04048814cedfa2aa8ff2dcc/storage/raft_transport.go#L56)):*
> ```Go
> type RaftMessageHandler interface {
> 	HandleRaftRequest(*RaftMessageRequest) *roachpb.Error
> 	HandleRaftResponse(*RaftMessageResponse) *roachpb.Error
> ```

what's the point of having this return a structured error? the structuredness is immediately discarded.

there's a fundamental asymmetry here, i don't think we need to make these methods look alike.

---

*[storage/raft_transport_test.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZ3JbvoL6xT6otD0n:-KOxZ3JbvoL6xT6otD0o:bdjj350) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport_test.go#L76)):*
> ```Go
> 
> func (s channelServer) HandleRaftResponse(resp *storage.RaftMessageResponse) error {
> 	log.Fatalf(context.Background(), "unexpected raft response: %s", resp)
> ```

`log.Fatal` is pretty nasty in tests. can this return an error instead?

---

*[storage/store.go, line 2499 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZJqvHihngvJfkbJt:-KOxZJqvHihngvJfkbJu:bmpo3ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/store.go#L2499)):*
> ```Go
> }
> 
> // HandleRaftResponse handles responses messagaes from the raft
> ```

this comment needs a once-over

---

*[storage/store.go, line 2501 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZPrU48MJfFp4CJFJ:-KOxZPrU48MJfFp4CJFK:b-xrfkyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/store.go#L2501)):*
> ```Go
> // HandleRaftResponse handles responses messagaes from the raft
> // transport. It requires that s.processRaftMu and s.mu are not held.
> func (s *Store) HandleRaftResponse(resp *RaftMessageResponse) error {
> ```

add a context argument. You can grab the context using `stream.Context()`

---

*[storage/store.go, line 2513 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZrxBiIyUkNI5HZWs:-KOxZrxBiIyUkNI5HZWt:b7j7hr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/store.go#L2513)):*
> ```Go
> 					log.Infof(ctx, "%s: replica %s too old, adding to replica GC queue", rep, resp.ToReplica)
> 
> 					if err := s.replicaGCQueue.Add(rep, 1.0); err != nil {
> ```

does this need to be in a task?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-:-KOxaL4FhU0WaGxJdZEm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.proto, line 198 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOx_GWSUV260oc1i0Rt:-KOykIDFFuIp73tVNbKP:bdw4t0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/18cc7fb935fee525e04048814cedfa2aa8ff2dcc/roachpb/errors.proto#L198)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you make this comment more specific by discussing its sender and recipient?
</details>
@tamir The comment seems fine to me. There is no harm in verbosely stating the change you think would be helpful. 

---

*[storage/raft_transport.go, line 367 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOylZ_GEKkpC7lABuz0:-KOylZ_GEKkpC7lABuz1:b-qlus78) ([raw file](https://github.com/cockroachdb/cockroach/blob/b573b1ab7181e6e6d8afc809cf57d9dce5d90dbe/storage/raft_transport.go#L367)):*
> ```Go
> 				if err := t.processQueue(ch, conn); err != nil && !grpcutil.IsClosedConnection(err) {
> 					log.Warningf(context.TODO(),
> 						"range %s: outgoing raft transport stream to %s closed by the remote: %s",
> ```

`s/range %s/range=%d/g` for consistency with other logging.

---

*[storage/store.go, line 2502 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOymHvvK8rk2MAtP5yW:-KOymHvw6PIxstS_YVB1:b-wuggn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b573b1ab7181e6e6d8afc809cf57d9dce5d90dbe/storage/store.go#L2502)):*
> ```Go
> // transport. It requires that s.processRaftMu and s.mu are not held.
> func (s *Store) HandleRaftResponse(resp *RaftMessageResponse) *roachpb.Error {
> 	ctx := context.TODO()
> ```

There is a context in the response stream that you could pass in here.

---

*[storage/store.go, line 2529 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOymU1dzak-3dKMFeK5:-KOymU1egBJC9NcxZh1f:b-7ha32m) ([raw file](https://github.com/cockroachdb/cockroach/blob/b573b1ab7181e6e6d8afc809cf57d9dce5d90dbe/storage/store.go#L2529)):*
> ```Go
> 		log.Infof(ctx, "%s: got unknown raft response type %T: %s", s, val, val)
> 	}
> 	return nil
> ```

You're never returning an error, so perhaps get rid of the return value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-:-KOymZWIy7jIa8IfykQr:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.proto, line 198 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOx_GWSUV260oc1i0Rt:-KOyvLgGV2g96BldbUCw:b2q734j) ([raw file](https://github.com/cockroachdb/cockroach/blob/18cc7fb935fee525e04048814cedfa2aa8ff2dcc/roachpb/errors.proto#L198)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> @tamir The comment seems fine to me. There is no harm in verbosely stating the change you think would be helpful. 
</details>
I hadn't formulated the comment, but I suppose it'd be something like:

ReplicaTooOldError is sent in response to a raft message when the recipient of the raft message believes the sender of the raft message to have been removed from the raft group (i.e. the range).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on storage/raft.proto:
```diff
@@ -35,9 +36,28 @@ message RaftMessageRequest {
 
-// RaftMessageResponse is an empty message returned by raft RPCs. If a
-// response is needed it will be sent as a separate message.
+message RaftMessageResponseUnion {
```
This is a strange construction. What else do you envision being added to this union? Why not just make the error optional in the `RaftMessageResponse`?

spencerkimball commented on storage/store.go:
```diff
@@ -2482,20 +2482,53 @@ func (s *Store) handleRaftMessage(req *RaftMessageRequest) error {
 	return nil
 }
 
+// HandleRaftResponse handles responses messagaes from the raft
```
s/responses messagaes/response messages/

bdarnell committed:
storage: Clean up RaftMessageHandler interface

Remove unused error return, add context, update comments

bdarnell commented:
Review status: 10 of 16 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[roachpb/errors.proto, line 198 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOx_GWSUV260oc1i0Rt:-KP-ksIE3lr1yQ82xtx6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18cc7fb935fee525e04048814cedfa2aa8ff2dcc/roachpb/errors.proto#L198)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I hadn't formulated the comment, but I suppose it'd be something like:
> 
> ReplicaTooOldError is sent in response to a raft message when the recipient of the raft message believes the sender of the raft message to have been removed from the raft group (i.e. the range).
</details>
Done.

---

*[storage/raft.proto, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxXpgohG1e8X_kU4sT:-KP-l9WiPyWyD8rPq0m9:b-4orcge) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft.proto#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what are you anticipating here?
</details>
See reply on spencer's comment.

---

*[storage/raft.proto, line 39 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOvg48WKjuG6MigIM2R-r3-39:-KP-ktw3Ly_1lI25Phhp:bhr4nug) ([raw file](https://github.com/cockroachdb/cockroach/blob/b573b1ab7181e6e6d8afc809cf57d9dce5d90dbe/storage/raft.proto#L39)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> This is a strange construction. What else do you envision being added to this union? Why not just make the error optional in the `RaftMessageResponse`?
</details>
I'll use it for snapshot acknowledgements, although that is admittedly a stopgap that will be removed when we have streaming snapshots. We may also use it for proposal acks and/or flow control, depending on how those fit in between our code and etcd/raft.

---

*[storage/raft_transport.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxY81qJuN5YdMcB1No:-KP-lBad7Wv9kJaJfLFi:b-p4ybu) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L53)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this appears to be untrue of `HandleRaftRequest`.
</details>
Oops, fixed.

---

*[storage/raft_transport.go, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYFm7Z9kPHufQJq46:-KP-lG_n4Vk6S_E-xG7L:b-5zhd7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L178)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think this wants to send the "node is stopping" error.
</details>
This is the "clean shutdown" case; we shouldn't be spamming the logs for things like this without good reason. 

---

*[storage/raft_transport.go, line 271 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYR-1ZqzuqenItYHV:-KP-lQ2w0vfGIGdr6HjH:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L271)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Why not `stream.Recv()`?
</details>
Changed.

---

*[storage/raft_transport.go, line 369 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYpH30CJ_h5aB2Q3Y:-KP-lRijCuzGyUv2ztJ7:b7cyt9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L369)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you don't want to log this on nil errors?
</details>
No; clean shutdowns should be quiet. 

---

*[storage/raft_transport.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYwNph6mK10dkdMYs:-KP-lZBCGDEhXdsUdLp4:b-chz8h0) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L379)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> perhaps `sendAsync` wants to return an error.
</details>
Maybe, but not today. The error here is just the result of RunTask, so it's not an interesting error (it's just an `error` to let us use `errcheck`). 

---

*[storage/raft_transport.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOx_e_YECMl5EAZPFDa:-KP-lCgQbmAYk8FxI7bz:bsob732) ([raw file](https://github.com/cockroachdb/cockroach/blob/18cc7fb935fee525e04048814cedfa2aa8ff2dcc/storage/raft_transport.go#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's the point of having this return a structured error? the structuredness is immediately discarded.
> 
> there's a fundamental asymmetry here, i don't think we need to make these methods look alike.
</details>
Removed the return value; we don't use it. 

---

*[storage/raft_transport.go, line 367 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOylZ_GEKkpC7lABuz0:-KP-lW5wDovQF1oDiiCB:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b573b1ab7181e6e6d8afc809cf57d9dce5d90dbe/storage/raft_transport.go#L367)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/range %s/range=%d/g` for consistency with other logging.
</details>
Done

---

*[storage/raft_transport_test.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZ3JbvoL6xT6otD0n:-KP-lmcvNdrZj78JPrIk:b-g1seub) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport_test.go#L76)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `log.Fatal` is pretty nasty in tests. can this return an error instead?
</details>
Unfortunately there's no good way for an error to get back to anything that could fail the test, so we have to fatal/panic (or do a lot of plumbing)

---

*[storage/store.go, line 2501 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZPrU48MJfFp4CJFJ:-KP-lw6uqjuwC44Wftpg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/store.go#L2501)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> add a context argument. You can grab the context using `stream.Context()`
</details>
Done.

---

*[storage/store.go, line 2513 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZrxBiIyUkNI5HZWs:-KP-lxc4Q4eF6LglDRAk:b-sfkql0) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/store.go#L2513)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> does this need to be in a task?
</details>
It was in the old code, so I'm doing the same here. 

---

*[storage/store.go, line 2529 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOymU1dzak-3dKMFeK5:-KP-m0owP08ZmT219NbJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b573b1ab7181e6e6d8afc809cf57d9dce5d90dbe/storage/store.go#L2529)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> You're never returning an error, so perhaps get rid of the return value.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.proto, line 198 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOx_GWSUV260oc1i0Rt:-KP-lO6ooBagK4KJ0072:b-h0utdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/18cc7fb935fee525e04048814cedfa2aa8ff2dcc/roachpb/errors.proto#L198)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I hadn't formulated the comment, but I suppose it'd be something like:
> 
> ReplicaTooOldError is sent in response to a raft message when the recipient of the raft message believes the sender of the raft message to have been removed from the raft group (i.e. the range).
</details>
end of the sentence seems to have been lost.

---

*[storage/raft_transport.go, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYFm7Z9kPHufQJq46:-KP-m9En-jtRajDj-o-i:b-9ow9yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L178)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is the "clean shutdown" case; we shouldn't be spamming the logs for things like this without good reason. 
</details>
Yeah, it's the clean shutdown as you say. It's really only relevant in tests. I'd err on the side of logging it in case it's relevant (again, test only), but I don't feel strongly.

---

*[storage/raft_transport.go, line 369 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYpH30CJ_h5aB2Q3Y:-KP-lssuThdNufkkfAN6:bxd97k3) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L369)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> No; clean shutdowns should be quiet. 
</details>
I believe the error will be `nil` if the remote returns `nil` from its RaftMessage handler.

---

*[storage/raft_transport.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYwNph6mK10dkdMYs:-KP-mgqYDn-kPNe5AF5R:b-ifn5h0) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L379)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Maybe, but not today. The error here is just the result of RunTask, so it's not an interesting error (it's just an `error` to let us use `errcheck`). 
</details>
OK. As above, it may be useful to see this go to no-op during shutdown when investigating future test failures.

---

*[storage/raft_transport_test.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZ3JbvoL6xT6otD0n:-KP-mzSJDOvIt-Yvw5SK:b2s6qdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport_test.go#L76)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Unfortunately there's no good way for an error to get back to anything that could fail the test, so we have to fatal/panic (or do a lot of plumbing)
</details>
You could pass the `*testing.T` to `newChannelServer`, and then `s.t.Error` here (`Fatal` is probably unsafe). `newChannelServer` only has 2 callers - one has the `*testing.T` in scope and the other would require you to pass the `*testing.T` to `(*raftTransportTestContext).ListenStore` which has 3 callers, all of which have the `*testing.T` in scope.

---

*[storage/store.go, line 2513 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZrxBiIyUkNI5HZWs:-KP-oO8-2aflOIJRIQ-8:b-bjcobh) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/store.go#L2513)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It was in the old code, so I'm doing the same here. 
</details>
Seems misleading - I don't see anything in `Add` that depends on other services.

Can you add a TODO?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxYFm7Z9kPHufQJq46:-KPAwNE7qWZAZXwDB4NL:b2mn0m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport.go#L178)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yeah, it's the clean shutdown as you say. It's really only relevant in tests. I'd err on the side of logging it in case it's relevant (again, test only), but I don't feel strongly.
</details>
Clean shutdown is not just test only - I'm more concerned about log spamminess in production, where rolling restarts shouldn't result in lots of log messages across the fleet. 

---

*[storage/raft_transport_test.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZ3JbvoL6xT6otD0n:-KPAw3eMd4e1Z15pRvS7:bv3d3jl) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/raft_transport_test.go#L76)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> You could pass the `*testing.T` to `newChannelServer`, and then `s.t.Error` here (`Fatal` is probably unsafe). `newChannelServer` only has 2 callers - one has the `*testing.T` in scope and the other would require you to pass the `*testing.T` to `(*raftTransportTestContext).ListenStore` which has 3 callers, all of which have the `*testing.T` in scope.
</details>
Yeah, we can do that when and if we have a test that uses channelServer and has a reasonable likelihood of transport errors. 

---

*[storage/store.go, line 2513 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KOxZrxBiIyUkNI5HZWs:-KPAwAJrtEDc52kvavdf:b-djtkjj) ([raw file](https://github.com/cockroachdb/cockroach/blob/984f21b62f9d2afa9c473ed1b5aabbe5755b4da1/storage/store.go#L2513)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Seems misleading - I don't see anything in `Add` that depends on other services.
> 
> Can you add a TODO?
</details>
Seems better to assume that the old code has a reason for being the way it is and not waste someone's time on reinvestigating things where the potential gain is so small. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Impact of this change on the beta cluster: Number of goroutines per node down from 25-40K to 150-200. Go memory usage (total across all nodes) down from 17GB to 3.8GB. Cgo memory usage up from 23 to 30GB (not sure what's going on here). Much less Go runtime stuff in the CPU profile now, although we're still seeing a lot of time spent growing/reallocating stacks (unfortunately this doesn't tell us which functions are requiring the stack to grow). Despite all of this, the cluster is still not available, with similar log messages as before.

arjunravinarayan commented:
*[storage/raft_transport.go, line 89 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495#-KPF5HV7oRU61IB_NJNn:-KPF5HV7oRU61IB_NJNo:b-k58owj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1fa43eec6d440cbea8478673dab5a34f6ca6bb/storage/raft_transport.go#L89)):*
> ```Go
> // queued on a per-destination queue before being asynchronously delivered.
> //
> // Callers are required to construct a RaftSender before being able to
> ```

Isn't this comment now obsolete?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8495)*
<!-- Sent from Reviewable.io -->
